horovod/run/common/util/safe_shell_exec.py

```diff
@@ -69,75 +74,104 @@ def prepend_context(line, rank, prefix):
-            if not isinstance(text, str):
-                text = text.decode('utf-8')
-            if not text:
-                break
+    line_buffer = ''
+    while True:
+        text = os.read(src_stream.fileno(), 1000)
+        if not isinstance(text, str):
+            text = text.decode('utf-8')
+        if not text or len(text) == 0:
+            break
 
-            for line in re.split('([\r\n])', text):
-                line_buffer += line
-                if line == '\r' or line == '\n':
-                    if index is not None:
-                        line_buffer = prepend_context(line_buffer, index, prefix)
+        for line in re.split('([\r\n])', text):
+            line_buffer += line
+            if line == '\r' or line == '\n':
+                if index is not None:
+                    line_buffer = prepend_context(line_buffer, index, prefix)
 
-                    dst_stream.write(line_buffer)
-                    dst_stream.flush()
-                    line_buffer = ''
+                dst_stream.write(line_buffer)
+                dst_stream.flush()
+                line_buffer = ''
 
-        # flush the line buffer if it is not empty
-        if len(line_buffer):
-            if index is not None:
-                line_buffer = prepend_context(line_buffer, index, prefix)
-            dst_stream.write(line_buffer)
-            dst_stream.flush()
+    # flush the line buffer if it is not empty
+    if len(line_buffer):
+        if index is not None:
+            line_buffer = prepend_context(line_buffer, index, prefix)
+        dst_stream.write(line_buffer)
+        dst_stream.flush()
 
+    src_stream.close()
 
-def execute(command, env=None, stdout=None, stderr=None, index=None, event=None):
-    # Make a pipe for the subprocess stdout/stderr.
-    (stdout_r, stdout_w) = os.pipe()
-    (stderr_r, stderr_w) = os.pipe()
 
-    # Make a pipe for notifying the child that parent has died.
-    (r, w) = os.pipe()
+def _exec_middleman(command, env, exit_event, stdout, stderr, rw):
+    stdout_r, stdout_w = stdout
+    stderr_r, stderr_w = stderr
+    r, w = rw
+
+    # Close unused file descriptors to enforce PIPE behavior.
+    stdout_r.close()
+    stderr_r.close()
+    w.close()
+    os.setsid()
 
-    middleman_pid = os.fork()
-    if middleman_pid == 0:
-        # Close unused file descriptors to enforce PIPE behavior.
-        os.close(w)
-        os.setsid()
+    executor_shell = subprocess.Popen(command, shell=True, env=env,
+                                      stdout=stdout_w, stderr=stderr_w)
 
-        executor_shell = subprocess.Popen(command, shell=True, env=env,
-                                          stdout=stdout_w, stderr=stderr_w)
+    def kill_executor_children_on_event():
+        exit_event.wait()
+        terminate_executor_shell_and_children(executor_shell.pid)
 
-        sigterm_received = threading.Event()
+    in_thread(kill_executor_children_on_event)
```

Comment:
can be replaced with `on_event(exit_event, terminate_executor_shell_and_children)`